### ISSUE TITLE
Feature Request: Dual Page mode

### ISSUE THREAD

**Issue by TAAPArthur** [nsxiv/nsxiv#35]
Summary: Planning to implement a new feature. Wanted to share my thoughts before starting in case anyone had concerns/ideas or was about to do the same thing. As always default behavior won't change.

One of my main use cases of sxiv would be for a manga reader, and one key feature for that is the ability to look at images side by side.

There are for cases that need to be supported. Plan to have an enum of them with the default by single-page view. Suppose you are viewing 2 images 1 and 2:
1. Left-to-right
|1|2|
2. Right-to-left
|2|1|

3. Left-to-right with offset
| |1|

4. Right-to-left with offset
|1| |

My general plan is change how images are loaded. Instead of loading an image, the next image will also be loaded. I'll create a temporary image with whose size of 2* MAX_WIDTH by MAX_HEIGHT and paste the 2 real images in it use the concatenated image for all subsequent ops. If anything needs to know file path or other image metadata, the first image's info will be used.

I'm not yet familiar with imlib, but I think the above is feasible and not too invasive. For keybindings/cli flags just need to an ability to select the mode. Also plan to a multiplier (defaulting to 1) that will adjust all the file increment ops. For example in mode 1, a sane multiplier would be 2, so if using a keybinding to go the the next file, file_idx will be incremented by 2 instead.

Possible extensions would be extending this from a 1x2 grid to a user specified NxM grid. The 2x1 case may be interesting for vertical comic strips. And at that point it would probably be easier to just support a NxM grid.

Abandoned ideas
Repurposing thumbnail mode. Thumbnails are basically already a grid. Changing the img size and limiting the number shown and removing the padding would yield something close to the end result. Decided against it mainly because of the loss of ability to zoom/pan the the unneeded ability to select individual images

Having an array of images instead of combining into one giant image. Thought this would be too invasive and would require a lot of refactoring since most functions assume a single image.

**Comment by eylles** [comment]
hmmm i can see the use for showing 2 images side by side, however a lot of this if not shy of all can be done in a wrapper as combining images is something image magic can do, the question pertaining rn is how do you usually read manga, as in, if you download it from a service like mangadex or fanfox (which i do use) then pages that are meant to be read with the booklet wide open ie side by side are already concatenated into a single image, at least on those 2, i don't know if there's some service that doesn't concatenate those kinds of pages side by side as a file but instead does it dynamically on the client side.

as said 2 pane image viewing can be a neat feature to add, but this case sounds like a lot of work for a built in feature.

however as a wrapper something interesting happens as we get 2 cases

case 1: purely wrapper handled, this works even with muennich's last release
you can start either with an existing gallery on disk or a remote gallery that has to be downloaded (on which case the wraper may need to include the nsxiv-url code to download the images into a tmp dir)
then run `identify -format '%w %h\n' image` on every image inside directory, this will directly output the dimensions of the image passed as argument, and then a comparison can be made to find any double sized image, if no double sized image has been found then an image magick command can run to create a black page, then horizontally append the black image to the first page, the last page to the black image, and then every pair of images in between, perhaps have a check to determine if the number of images in the dir is odd or even, if it is odd the last page doesn't get the black image appended to it, write every appended image pair to a temporary directory
finally run sxiv on the temporary directory of appended images
trap sxiv exit to clean up the tmp dir

case 2: if sxiv supported piping
the wrapper would be simplified as there would be no need to write the appended images to a tmp dir as they could simply be piped to sxiv

again not saying a double pane mode won't be a cool feature, but for this specific case the wrapper can suffice (please add it to the wiki)

**Comment by TAAPArthur** [comment]
I use [some fancy aggregations script ](https://github.com/TAAPArthur/amt). Basically it will download desired manga from a wide variety of legal sources, write to disk and then open them in a specified image viewer. One such [provider](https://mangaplus.shueisha.co.jp/updates) stores dual-page images as 2 separate images. Another downside of the wrapper is not all sources use the same offset. Some may expect mode 1 (no offset) and others mode 2 (with offset). Then there is also the possibility of vertical manga where horizontal dual page mode doesn't make since or may even be worse than normal

Currently I use zathura and have keybindings to toggle between the 2. The problem with it is that it determines the img size to fit to based on the first/largest image so if that isn't representative the other images aren't the right size.

The wrapper is a nice idea. My main hesitation is not being certain about the offset to start with and not always wanting to be in dual-page mode. The majority of time, I get N pages of all equal sizes and generally only care about dual-page when I need it to view the entirety of the image. Could probably manage to make a good guess and could have a command to restart and rerun the wrapper and have negate the offset and jump to where I left off. I feel like just modifying the temporary dir and having nsxiv auto-reload wouldn't easily preserve my current position

For case 2 (piping), it is a bit of a shame to lose the file name info. Right now I have each img name contains the page number. Also having all the images loaded in memory isn't ideal, although probably not a real problem

**Comment by eylles** [comment]
I just realized something, this intended usecase will benefit from another of the intended features planed, the dmenu mode, @N-R-K said he will take a look at it and try to open a draft for a better dmenu mode tomorrow.
With the addition of a dmenu mode (altho it can work now with the -o flag but not as seamless) the agregator script can call sxiv such that std out will be stored on a var,  if sxiv exits and the var is empty then the manga chapter can be marked as read full, however if the var has the name of an image file, then that image file is the last page read.

**Comment by N-R-K** [comment]
A couple thoughts:

* Is there any external program that can do splits, something like [suckless's tabbed](https://tools.suckless.org/tabbed) ? <br> If so, then we should try to make sxiv work with that instead, since we already support `XEmbed` protocol. *Extensibility* should be higher priority over hard-coding one specific feature.

* If there isn't, should we instead try to add split functionality to tabbed, maybe try to get it into mainline tabbed instead of having it be a community patch. <br> While I'm aware that this might be out of scope for us, this would serve far more usecases and many more users than hardcoding splits/dual pane into sxiv itself.

If the above two options aren't possible, then we should still explore more possibilities on how we can _extend_ sxiv so that dual pane, but also other features/scripting possibilities can be added before resorting to hard-coding dual pane into sxiv itself.

- - -

> try to open a draft for a better dmenu mode tomorrow.

Not today, but soon. I need to get a bit more familiar with the codebase to figure out what the optimal way would be to add such feature.

**Comment by eylles** [comment]
After some thinking over and considering nrk's suggestion of using a tool similar to tabbed but that can do splits (like a tmux for x11 programs) or adding that to tabbed (patch or mainline), i think the split side by side view doesn't compete to nsxiv's scope directly and is left to an outside tool but doing so raises a problem, how would you control 2 sxiv instances simultaneously?

Now i know of xdo, but that can get messy real fast.

Now to an example pertaining the side by side page manga reader use case, how would you pass the "page" on both running sxiv instances? (Lets ignore xdo for a while) answer, you would have to do it manually... not ideal
However i tought of a quite unix answer to this problem, unix sockets 

From a high level view a unix socket isn't more than a file, to which you can `echo "text" > program-socket` and then the program that the socket belongs to will read the text from the socket, parse it, and if it is a valid function with a valid argument execute the function.

Now i'm not sure if the proper way to do unix sockets is every instance reads from a single socket or each instance has it's own socket, but single socket would get real messy real fast, so let's say each sxiv instance creates it's own socket called "title_prefix"-socket_"instance_num(this can be the PID)", for the manga reader case then stuff gets simplified as you don't even need tabbed anymore, all you need is a tiling window manager to tile the instances, left "pane" will have the title "sxiv L pane" and the right will be "sxiv R pane", with this then you can have a script that recieves an argument, let's say sxiv_mangareader_ctl.sh and all that script has to do when you run it with the `next` argument is 
`echo "i_navigate, +1" > sxiv\ L\ pane-socket_0 ;
echo "i_navigate, +2" > sxiv\ R\ pane-socket_0`
Which would properly move the pages on both instances as expected.

This approach also has the long term advantage of added extensibility and scriptability through unix sockets, with that we could then add a function to add images to an already running sxiv instance and not have to worry about having a built-in interface for such function other than the unix socket, so that then you can have something like this in your key handler for when you rename an image
``` sh
N) new_name=$(dmenu)
      mv "$file" ./"$new_name"
      echo "add_image, ./${new_name}" > sxiv-socket
      ;;
```

Now onto the concern of this usecase breaking for floating/stacking window managers that would be outside of nsxiv's scope and the tiling would have to be managed in a wm per wm case, or with a external tool, an hypotetical `xspit` (that may as well be a patched `tabbed`), for which we would have to either write such program (again outside nsxiv's scope, altho the repo can be under the same organization) or write a patch for tabbed.

At least this is my take and for the time being seems the easier and simpler route to go along being the one that offers more long term benefits.

**Comment by N-R-K** [comment]
Being able to control a running instance of sxiv from the cli or through a script will open up a huge amount of possibilities when it comes to scripting and extensibility in general.

I can easily see how useful this would be, although (I presume) this will require substantial work the long term benefit of it should justify it.

**Comment by TAAPArthur** [comment]
@eylles I think I really like that xsplit idea, but I'm thinking of it slightly differently.
We have a program xsplit that takes an array of programs + args. It creates dummy window and spawns all of its arguments such that they embeded themselves in the dummy window. The dummy window is what the WM gives focus to/and what the user will send key events to and xsplit will forward all events to each window.

So for syntax, it'd be something like `xsplit "nsxiv --offset 0 $MANGA_DIR" "nsxiv --offset 1 $MANGA_DIR"` to show the dual pages. The only assumptions is the the program in question doesn't ignore user generated event. For how the tiling would work, don't think it would be unreasonable to offer a simple grid layout in xplit, and it also isn't impossible to tell a WM what it should consider a root window ie could have the full flexibility of a tiling window manager (including keybindings) for each xplit instance if one so wished. I don't know of any WM that does so, but it should be easy enough to do. Of course there exists smaller tiling/keybindings programs that could be used/repurposed too.

Only changes requested of nsixv would be adding the multiplier (cli option) for file increments so the next page keybinding would change all instances by N pages. And probably the ability to go past the first/last file in the list and show a blank page. And would need to control image centering. If an image has been scaled so its height fits the window, then it would probably need to be left/right aligned (cli option). Any objection to the changes suggested in this paragraph? If not, any preference for this "xsplit" to be in this repo/org/wiki as it is a helper tool or just hosted on my own account? I don't expect to be very large (<200 if I try to bloat it) but it also won't be really specific to nsxiv.

>Being able to control a running instance of sxiv from the cli or through a script will open up a huge amount of possibilities when it comes to scripting and extensibility in general.

Also for the record some alternatives to sockets would be having a single nsxiv instance that spawns multiple windows or controlling nsxiv by X-client-message. There are trade offs, but think they are at least worth mentioning.

**Comment by N-R-K** [comment]
> Also for the record some alternatives to sockets would be having a single nsxiv instance that spawns multiple windows or controlling nsxiv by X-client-message. There are trade offs, but think they are at least worth mentioning.

Maybe fifos too.

> Only changes requested of nsixv would be adding the multiplier (cli option) for file increments so the next page keybinding would change all instances by N pages.

Assuming we implement some way to communicate with sxiv to tell it to go to the next image, changing image on all instance should be done with a wrapper instead.

> If not, any preference for this "xsplit" to be in this repo/org/wiki as it is a helper tool or just hosted on my own account? I don't expect to be very large (<200 if I try to bloat it) but it also won't be really specific to nsxiv.

No strong preference here. Although I think implementing split to [tabbed](https://tools.suckless.org/tabbed/) would be better. But if you want to roll your own, that's fine too.

> Currently I use zathura and have keybindings to toggle between the 2. The problem with it is that it determines the img size to fit to based on the first/largest image so if that isn't representative the other images aren't the right size.

Well... for your use-case at least, don't you think it might be easier to just patch zathura?

**Comment by TAAPArthur** [comment]
> Well... for your use-case at least, don't you think it might be easier to just patch zathura?

Honestly it isn't. Found the code much more complex and harder to hack on. The issue is noted on their gitlab. Manually adjusting all the image sizes so they match would be easier. However using a raw image viewer has benefits over bundling all the images into a zip. This project also has nice features like better control of the window title. That said I'm not married to the idea of using nsxiv for this use case. Just wanted to toss some ideas around. 

>>Only changes requested of nsixv would be adding the multiplier (cli option) for file increments so the next page keybinding would change all instances by N pages.

>Assuming we implement some way to communicate with sxiv to tell it to go to the next image, changing image on all instance should be done with a wrapper instead.

Just to make sure we are on the same page. I meant for the external script to tell all instances to go to the "next" page and have all instances know what offset to use. Instead of having the script tell all instances to go the the next page twice or to jump to an relative/abs position. I'm not completely against such an idea but think it would make the wrapper script more complicated. However, it does seem like overkill for a config option if this is the only usecase. This is one reason why https://github.com/nsxiv/nsxiv/issues/50 would be handy. Would let people do their own nice thing from config.h rather needing to patch

**Comment by eylles** [comment]
thinking about the hypotetical xsplit i still can't decide if ti would be better off as a part of tabbed or as an idependient program, as i see both cases having advantages and disadvantages.

advantages of xsplit as independent program:
* no need for tabbed
* xsplit can be embedded into tabbed
* a running tabbed can be embedded into xsplit

advantages of xsplit as part of tabbed:
* can convert tabs into splits and splits into tabs

**Comment by N-R-K** [comment]
> * xsplit can be embedded into tabbed
> 
> * a running tabbed can be embedded into xsplit

Can it? I'm not sure if you can do nested embedding, but if the XEmbed protocol allows that, then I agree, it would be better off as a separate program.

**Comment by eylles** [comment]
well, we can't know unless we try it but i see both having their advantages, i just tried embedding an instance of tabbed inside another and it couldn't be done so i guess that's a no, which may mean those 2 advantages of xsplit being it's own thing don't count.

**Comment by TAAPArthur** [comment]
@eylles I'm pretty sure you can embed an embedded window. I didn't think there was a limit. At is core the XEmbed protocol is just re-parenting and given how many WMs and programs do so I just assume there isn't such a small limit. But haven't actually done it so could be overlooking something, but I'd bet that's a bug in tabbed.

**Comment by TAAPArthur** [comment]
So quick search for "xsplit" came up with results, so I'm calling this new thing "gridded".
I have a draft of if. It can be nested in tabbed and within itself. Don't think it can nest tabbed, because tabbed doesn't seem to support an option to specify the window to embed into. Should be simple to augment tabbed if anyone cared.

Right now the basic syntax is 
`./gridded -c2 './nsxiv -e $GRIDDED_PARENT_WIN! ~/Pictures/' './nsxiv -e $GRIDDED_PARENT_WIN! -n2 ~/Pictures/'` where `GRIDDED_PARENT_WIN` is an env var gridded sets to denote its window and the `!` suffix is used to indicate that nsxiv should listen for events on the window it embeds itself in instead of its own window. The way this is implemented breaks support for mouse events unfortunately. Since it isn't fully compatible with all existing features probably won't get mainlined.

So right now I can see 2 instances side by side and any key pressed is seen by both. So pressing "space" goes to the next image for both instances.

Next steps would be to add keybindings to `gridded` to swap the number of rows/cols to from a horizontal dual-page to vertical dual page and then to toggle a full-screen view (1x1 grid) and back. 
Then for nsxiv 
- need to figure out the best way to determine how many pages/increment by I want to increment. Ideally "space" would do the right thing and increment by R*C for any R,C sized grid. May just set a property and have the increment function read it once #50 lands. Should be easy enough to do from a config file
- Also need special handling for going past the first/last page.
- And then left/right image alignment

Going to try and get these last 2 into nsxiv at least.

**Comment by eylles** [comment]
there are 3 ways to do this now: 
1. the "lazy" way would be to simply have the left instance of nsxiv load all the odd images, and the right instance load all the even images.
2. the future feature way by adding unix sockets to control an already running instance, which you can start the 2 nsxiv instances like you already did and would allow to do something like this `echo "i_navigate, +2" > nsxiv_left_socket ; echo "i_navigate, +2" > nsxiv_right_socket` to show the next image and something like this `echo "i_navigate, -2" > nsxiv_left_socket ; echo "i_navigate, -2" > nsxiv_right_socket` to show the previous image.
3. the "hacky" which would be to use an xdo script to send a keybing to both instances, like 2n or the next image on both or 2p for the previous (remember nsxiv can recieve numerical prefixes to most keybinds)

**Comment by N-R-K** [comment]
Sxiv can reload the image on change. So we one do something like this.

```sh
cp 1.png left.png
cp 2.png right.png

nsxiv left.png
nsxiv right.png

cp -i 2.png left.png
cp -i 3.png right.png
```

I'm perfectly aware how hacky this is, but it might take care of changing images in running sxiv instance.

**Comment by eylles** [comment]
Yeh it is horribly hacky, also it presents a problem to taap's usecase as he want's to be able to "save" the name of the current image viewed (with the wrapper) so that if he quits before reading all the manga chapter and then later relaunches the script with said manga chapter then he can pick up fromwhere he left automatically.

**Comment by N-R-K** [comment]
That can very easily be done too. Also if you're concerned about disk weardown, then you can easily use `ramdisk` for that.

I'm not saying this is the correct solution, or saying that we shouldn't add a more robust way to communicate with a running sxiv instance. But I'm just throwing the idea around.

**Comment by eylles** [comment]
perhaps if this feature request will depends on the prototype gridded and on a possible future feature we should add a issue about such a future feature (unix sockets, or any other way to comunicate to an already running instance) and start a repo for the prototype program, question now is if having gridded be a repo under the nsxiv organiztion.

**Comment by N-R-K** [comment]
> having gridded be a repo under the nsxiv organiztion

Since taap is going to be doing the bulk of the work, it's completely up to him. If he wants it under nsxiv org, maybe for visibility reason, I at least won't object. But if he want's to have more control over the repo then he can have it under his account too.

**Comment by TAAPArthur** [comment]
> Yeh it is horribly hacky, also it presents a problem to taap's usecase as he want's to be able to "save" the name of the current image viewed (with the wrapper) so that if he quits before reading all the manga chapter and then later relaunches the script with said manga chapter then he can pick up fromwhere he left automatically.

This isn't a requirement from me. I only want to "resume" where I left off when switching from single to dual or vice versa. Once I shutdown nsxiv (and not restart), it can forget everything.

>I'm perfectly aware how hacky this is, but it might take care of changing images in running sxiv instance.

It is ideas like that that I personally am looking for. Maybe just cause it is cool that it actually works. I was leaning towards having gridded be able to cycle images along rows/cols.

> question now is if having gridded be a repo under the nsxiv organiztion.

I don't have a preference either way. Was leaning towards putting it under my account and linking to it in the wiki mainly because the only relation it has to nsxiv is the somewhat niche request. If there is desire to have it under this umbrella, I don't think I have reason to object

**Comment by TAAPArthur** [comment]
So implemented [gridded](https://github.com/TAAPArthur/gridded) to a passable level. The only real changes need to nsxiv was the ability to listen to the parent window and the ability to right/left align images instead of just centering them.

Basically I just spawn two instances of nsxiv in gridded. Gridded can flip the windows to effectually change reading direction and it can gridded toggle between showing both or only 1. The same keybinding used to toggle is also visible for nsxiv and is bound to a function to track how the default page increment. The incrementing function is also smart enough to not go past the end/beginning. It does make assumptions about the parity of the pages which is fine for my usecase. The main problem for me is if there is an invalid image, the 2 windows will get out of sync. Same is true for an absolute jump. Thinking of just adding a binding to get the windows to resync themselves based on the startup option.

Since my goal has been met I'm closing the issue

**Comment by eylles** [comment]
>The only real changes need to nsxiv was the ability to listen to the parent window and the ability to right/left align images instead of just centering them.

hmm left or right align can be a pr similar to the scale mode Fit

on listening to the parent window i think a more generic ipc would be better in the long run

**Comment by TAAPArthur** [comment]
>hmm left or right align can be a pr similar to the scale mode Fit

Sure I'll submit the PR.

>on listening to the parent window i think a more generic ipc would be better in the long run

A more generic solution would have been more work for less benefit for my usecase. Very literally needed all my keypresses to be forwarded to all windows. Thought of having a dedicated program manually forward them but the net result would have been the same. Also given that nsxiv can be fully controlled by keypresses, communicating with an instance by sending keypresses seems to a generic enough. If we don't want to go through the hassle of sending keyevents we could send client messages containing the keys we want to trigger or perhaps the function we want directly. I could also be missing the use case you are imaging.
